src/utils.ts

```diff
@@ -1811,48 +1896,104 @@ export class newText extends newShape {
-        if (splitDecimal.length > 1) { this.text = `${this.text}e${splitDecimal[1]}` };
+        while (decimal[decimal.length - 1] == "0") decimal = decimal.slice(0, -1);
+        if (decimal.length != 0) this.text = `${splitText[0]}.${decimal}`
+        else this.text = splitText[0];
+        if (splitDecimal.length > 1) this.text = `${this.text}e${splitDecimal[1]}`;
       }
     }
   }
 
-  private cutting_text(context: CanvasRenderingContext2D, maxWidth: number): string[] {
-    var words = this.text.split(' ');
-    var space_length = context.measureText(' ').width;
-    var cut_texts = [];
-    var i = 0;
-    var line_length = 0;
-    var line_text = '';
-    while (i < words.length) {
-      let word = words[i];
-      let word_length = context.measureText(word).width;
-      if (word_length >= maxWidth) {
-        if (line_text !== '') cut_texts.push(line_text);
-        line_length = 0;
-        line_text = '';
-        cut_texts.push(word);
-        i++;
-      } else {
-        if (line_length + word_length <= maxWidth) {
-          if (line_length !== 0) {
-            line_length = line_length + space_length;
-            line_text = line_text + ' ';
-          }
-          line_text = line_text + word;
-          line_length = line_length + word_length;
-          i++;
-        } else {
-          cut_texts.push(line_text);
-          line_length = 0;
-          line_text = '';
+  public format(context: CanvasRenderingContext2D): string[] {
+    let writtenText = [this.text];
+    let fontsize = this.fontsize ?? DEFAULT_FONTSIZE;
+    context.font = newText.buildFont(this.style, fontsize, this.font);
+    if (this.boundingBox.size.x) {
+      if (this.multiLine) [writtenText, fontsize] = this.multiLineSplit(fontsize, context);
+      else {
+        if (!this.fontsize || context.measureText(this.text).width > this.boundingBox.size.x) fontsize = this.automaticFontSize(context);
+      }
+    }
+    this.fontsize = fontsize;
+    context.font = newText.buildFont(this.style, fontsize, this.font);
+    this.height = writtenText.length * fontsize;
+    let longestRow = writtenText[0];
+    writtenText.forEach(row => { if (row.length > longestRow.length) longestRow = row });
+    this.width = context.measureText(longestRow).width;
+    return writtenText
+  }
+
+  public multiLineSplit(fontsize: number, context: CanvasRenderingContext2D): [string[], number] {
+    context.font = newText.buildFont(this.style, fontsize, this.font);
+    const oneRowLength = context.measureText(this.text).width;
+    if (oneRowLength <= this.boundingBox.size.x) return [[this.text], fontsize > this.boundingBox.size.y ? this.boundingBox.size.y : fontsize];
+    if (!this.boundingBox.size.y) return [this.fixedFontSplit(context), fontsize];
+    return this.autoFontSplit(fontsize, context);
+  }
+
+  private getWords(): string[] {
+    if (this.words) return this.words;
+    return this.splitInWords();
+  }
+
+  private splitInWords(): string[] {
```

Comment:
We already discussed it. I think your idea is good but since i need to catch each separators and glue it to the word, only for the first one, and then say that each additional separator between 2 words is a word. Then i want to write all chars considering previously built words, without loosing any one of them.

I think we could find a very performant method that would suit with what you propose but I think the it would be expensive for a small gain in perf